src/Plugin/resource/DataProvider/RateLimitedDataProvider.php

```diff
@@ -0,0 +1,183 @@
+<?php
+/**
+ * @file
+ * Contains Drupal\restful\Plugin\resource\DataProvider\RateLimitedDataProvider.
+ */
+
+namespace Drupal\restful\Plugin\resource\DataProvider;
+
+use Drupal\restful\Exception\ServerConfigurationException;
+use Drupal\restful\Plugin\resource\Field\ResourceFieldInterface;
+use Drupal\restful\RateLimit\RateLimitManagerInterface;
+
+class RateLimitedDataProvider implements DataProviderInterface {
```

Comment:
A data provider is a class that provides full crud on a given backend type (EMW, db_query, filesystem, plugin list, …). Therefore we need a _DataProvider_ for every backend type.

Additionally we have decorators. Those decorators are the candidates to be called _middle ware_. The decorators are _DataProviders_ that have an additional feature added to it (caching, …). The decorators can apply (add these features) to any _DataProvider_, providing the flexibility of horizontal inheritance.

There are different kinds of decorators:
  - Resource decorators: add functionality at the resource level. For instance: «increment the rate limit for ever processed request».
  - Data provider decorators: add functionality at the data provider level. For instance: «do not make a query for something we did before».
  - Resource field decorator: adds functionality to the object representing a single field. For instance: «add Field API capabilities to a generic public property».

As you can see, by using a common name in Computer Science as _decorator_ we can apply it freely throughout the code base. If we go by middle ware, we would need to specify the type of middleware and document how to create one. By calling it a decorator, developers can assume the it is a [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern) and know how to implement it.

Was this what you had in mind?